MINOR: Reduce in-memory copies of partition objects in onJoinComplete() and onJoinPrepare()

radai-rosenblatt opened the pull request:
`ConsumerCoordinator.onJoinPrepare()` currently makes multiple copies of the set of assigned partitions. We can let `subscriptions.assignedPartitions()` return a view of the underlying partition set, copy it only once and re-use the copied value.

radai-rosenblatt committed:
avoid useless copies of the set of assigned topic partitions

ijuma reviewed (commented):
Good catch. Would it make sense to change the method name to make it more obvious that a copy is returned? I haven't looked into it, so not sure either way.

satishd commented:
@ijuma Should not `PartitionStates.partitionSet()` return unmodifiable `Set` instead of creating a new `HashMap` instance for each invocation?

ijuma commented:
@satishd I was also wondering about that, but hadn't had a chance to look in more detail. It seems like can do that as the public KafkaConsumer methods do their own copy.

radai-rosenblatt commented:
i couldnt prove that the underlying collection wont be concurrently modified, so i chose to retain the current behaviour.

lindong28 commented:
I think the change is safe. Because if someone tries to modify `PartitionStates.partitionSet()` to return `map.keySet()` directory, the developer and the reviewer should verify that this does not cause problem for the caller.

@ijuma Personally I feel it is a bit verbose to name it e.g. `SubscriptionState.assignedPartitionsCopy()` and `PartitionStates.partitionSetCopy()`. Maybe we can just add Java doc for the method to clarify that the method should return a copy?

@radai-rosenblatt Can we also modify `KafkaConsumer.assignment()` so that it does not have to make an extra copy? The implementation of `PartitionStates.partitionSet()` shows clarify that this method returns a copy. But it is less clear for `SubscriptionState.assignedPartitions()`. Can we add a comment for `SubscriptionState.assignedPartitions()` that this method should return a copy?

@satishd Given that `PartitionStates.map` can be changes, it seems safer to still let `PartitionStates.partitionSet()` return a copy. I think we can make the method return unmodifiable Set if there is a way to prove that this won't cause problem.

ijuma commented:
@lindong28 the performance model is hard to understand if some methods return copies and others do not. If we can avoid copies altogether, that would be best. And users who need a copy can do that via a collection copy constructor.

lindong28 commented:
@ijuma Sure. Then let's update method name to specify that the returned object is a copy which won't be modified after it is returned.

radai-rosenblatt commented:
will do

radai-rosenblatt committed:
modify assignment-querying code-paths to return an unmodifiable live view of the subscription

radai-rosenblatt commented:
@ijuma @lindong28 I've rebased the patch and extended it to avoid copies entirely for the whole assignment() code path.
`KafkaConsumer.assignment()` was returning an immutable copy, so in that regard the semantics have not changed.
I've also added documentation to all methods involved clearly stating the return value is a live immutable view of the data.

what I'm still wondering though is whether `KafkaConsumer.assignment()` still needs to be done under a lock:

if the lock is really required, that means the method was designed to protect against concurrent modification of the set, which means a copy may still be required? (then again - the consumer is not known for being generally thread safe - so where is the race?)

on the other hand, if we are ok with returning an immutable **live** view to avoid copying - is the lock still required? if so - why?

radai-rosenblatt committed:
remove unused import

radai-rosenblatt committed:
remove copies from assignment handler path, re-introduce copy to assignments() calle, fix tests

radai-rosenblatt committed:
maintain API such that anything user-visible is a copy

ijuma commented:
@radai-rosenblatt Even though the consumer is not thread-safe, the existence of the heartbeat thread means that synchronization is needed sometimes. We had issues in the past when we didn't do copies like https://github.com/apache/kafka/commit/b31a7a9022bfb645cb61e6fbc0b22829108b3a3f. I'm not sure if the same applies to the `assignment` method, but it seems safest not return internal state to the user (similar to how you copied the collections before passing them to listeners). However, it also seems like a good idea for us to avoid unnecessary copies when using the collections internally (if safe).

radai-rosenblatt commented:
@ijuma @lindong28 patch has been changed such that `PartitionStates.partitionSet()` (which is in internals) is an unmodifiable view (and documented as such) and every user-visible collection (`assignment()` and rebalance listeners) is a copy.

lindong28 commented:
@radai-rosenblatt Thanks for the update. LGTM. Can you update the JIRA and PR description based on the latest implementation? Can we also change the title to make it sound a bit mild, e.g. "Reduce unnecessary copies of partitions for consumer.onJoinPrepare()".

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -452,9 +453,9 @@ protected void onJoinPrepare(int generation, String memberId) {
         // execute the user's callback before rebalance
         ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
-        log.info("Revoking previously assigned partitions {}", subscriptions.assignedPartitions());
+        Set<TopicPartition> revoked = new HashSet<>(subscriptions.assignedPartitions()); //copy
```
nits: can we put the comment above the line (as opposed to putting it at the end of the line) for consistency?

radai-rosenblatt replied:
:+1:

lindong28 commented:
Hey @radai-rosenblatt, can you see if we can address or discard the last comment and get this PR merged?

radai-rosenblatt commented:
@lindong28 - rebased, and cleaned up the comments

lindong28 commented:
Thanks for the patch. LGTM. Merged to trunk.
